pkgs/misc/uboot/default.nix

```diff
@@ -212,17 +206,116 @@ in {
       #
       # See https://github.com/afaerber/meson-tools/issues/3 for more
       # information.
-      '${buildPackages.armTrustedFirmwareTools}/bin/fiptool' create \
+      ${buildPackages.armTrustedFirmwareTools}/bin/fiptool create \
         --align 0x4000 \
-        --tb-fw '${firmwareBlobs}/gxb/bl30.bin' \
+        --tb-fw ${firmwareOdroidC2}/bl30.bin \
         --scp-fw bl301.padded.bin \
-        --soc-fw '${armTrustedFirmwareS905}/bl31.bin' \
+        --soc-fw ${armTrustedFirmwareS905}/bl31.bin \
         --nt-fw u-boot.bin \
         fip.bin
-      cat '${firmwareBlobs}/gxb/bl2.package' fip.bin > boot_new.bin
-      '${buildPackages.meson-tools}/bin/amlbootsig' boot_new.bin u-boot.img
-      dd if=u-boot.img of=u-boot.gxbb bs=512 skip=96
+      cat ${firmwareOdroidC2}/bl2.package fip.bin > boot_new.bin
+      ${buildPackages.meson-tools}/bin/amlbootsig boot_new.bin u-boot.img
+
+      dd if=u-boot.img of=u-boot.bin bs=512 skip=96
+    '';
+
+    filesToInstall = [
+      "u-boot.bin"
+      "${firmwareOdroidC2}/sd_fusing.sh" "${firmwareOdroidC2}/bl1.bin.hardkernel"
+    ];
+    extraMeta.platforms = ["aarch64-linux"];
+  };
+
+  ubootOdroidC4 = buildUBoot {
+    defconfig = "odroid-c4_defconfig";
+
+    # Fix eMMC/SD read issues on meson-sm1/meson-gx boards.
+    # Patches are queued for U-Boot 2021.01 release
+    patches = [
+      (fetchurl { #0001-mmc-meson-gx-move-arch-header-to-local-header.patch
+        url = "https://gitlab.denx.de/u-boot/custodians/u-boot-amlogic/-/commit/77863d43eb2b40319619bbb4f781270d8f027189.patch";
+        sha256 = "10nf0klspqmy15a8bb8ybhl53x11v1cqvh20i6bxyqm10zwlqbhl";
+      })
+      (fetchurl { #0002-mmc-meson-gx-change-clock-phase-value-on-SM1-SoCs.patch
+        url = "https://gitlab.denx.de/u-boot/custodians/u-boot-amlogic/-/commit/0dbb54eb3257c243c7968f967a6b183b1edb56c8.patch";
+        sha256 = "1xwj67h50gc32y7lrj7mh56s0zhwrd1vvxrpvhb2yhf1y17jf7rc";
+      })
+      (fetchurl { #0003-ARM-dts-meson-sm1-add-u-boot-specific-MMC-controller.patch
+        url = "https://gitlab.denx.de/u-boot/custodians/u-boot-amlogic/-/commit/c87eab81616d671a6004ffc95847bad21b7eb005.patch";
+        sha256 = "11l19vx8q2a8j95liv1nwkn9cjryxk9qgmj68ya4lr4j0c5qzh7c";
+      })
+    ];
+
+    # Gracefully omit build when run on native aarch64 machines. Also provide
+    # an empty u-boot.bin, which is safe in case someone will try to write it
+    # to SD or eMMC card.
+    # Due to absence of the "aml_encrypt_g12a" signing tool binary built for
+    # aarch64 - Odroid C4 must be cross-compiled on x86_64 machine.
+    dontBuild = stdenv.buildPlatform.isAarch64;
```

Comment:
It seems like it would be less confusing to just add an assertion (using `assertMsg`) explaining that it must be cross-compiled.